opds.py

```diff
@@ -79,12 +79,47 @@ class Annotator(object):
 
     opds_cache_field = Work.simple_opds_entry.name
 
-    @classmethod
-    def annotate_work_entry(cls, work, active_license_pool, edition,
+    def annotate_work_entry(self, work, active_license_pool, edition,
                             identifier, feed, entry, updated=None):
         """Make any custom modifications necessary to integrate this
         OPDS entry into the application's workflow.
+
+        :work: The Work whose OPDS entry is being annotated.
+        :active_license_pool: Of all the LicensePools associated with this
+           Work, the client has expressed interest in this one.
+        :edition: The Edition to use when associating bibliographic
+           metadata with this entry. You will probably not need to use
+           this, because bibliographic metadata was associated with
+           the entry when it was created.
+        :identifier: Of all the Identifiers associated with this
+           Work, the client has expressed interest in this one.
+        :param feed: An OPDSFeed -- the feed in which this entry will be
+           situated.
+        :param entry: An lxml Element object, the entry that will be added
+           to the feed.
         """
+
+        # First, try really hard to find an Identifier that we can
+        # use to make the <id> tag.
+        if not identifier:
+            if active_license_pool:
+                identifier = active_license_pool.identifier
+            elif edition:
+                identifier = edition.primary_identifier
+
+        if identifier:
+            entry.append(AtomFeed.id(identifier.urn))
```

Comment:
Previously the output of permalink_for() was used as the AtomFeed.id. But in the circulation manager we didn't actually want this behavior -- we always used the Identifier.urn because Annotator.permalink_for was always being called instead of LibraryAnnotator.permalink_for.

Some part of this change (I'm not sure what) made it so that the LibraryAnnotator was using its own permalink_for implementation, which revealed that we really didn't want the <atom:id> to be based on permalink_for(), we always want it to be Identifier.urn.